### ISSUE TITLE
How does the smartbugs adapt to different solidity versions?

### ISSUE THREAD

**Issue by JerryPW** [smartbugs/smartbugs#194]
Hi! Sorry to bother again. 
I've read some issues in this repository and tried by myself to find out that using ```solc-select``` seems not quite effective when dealing with different solidity version contracts. So, I'm curious about as facing multi-version contracts, how does the smartbugs implement to adapt it, since I always met ```solidity compilation failed``` message and I guess compilation version might be the problem. 
This is the first month I studied about solidity, therefore there might be some 'easy' question that I may find difficult. I apologize in advance if I asked some naive questions.

**Comment by JerryPW** [comment]
Here's an example of my problem. 
<img width="1493" alt="截屏2024-10-28 11 29 19" src="https://github.com/user-attachments/assets/41707b66-c989-4f75-982a-5eb4cc07f929">
This is a contract with reentrancy attack issue from https://bscscan.com/address/0x7ac55ac530f2C29659573Bde0700c6758D69e677#code . I tried to use ```sold-select use``` to change the solidity compiler from 0.4.23 to 0.6.6, noticing that all of the result turned out to be the same failing message. Therefore, I wonder if I want the compilation to work properly, what strategy should I use?

**Comment by gsalzer** [comment]
Thanks for providing an example illustrating the problem; it helped a lot in analyzing the issue.

- Local compiler versions do not affect SmartBugs, as it has it own mechanism to install the correct compiler for *Linux*, even if the host runs Windows. The compiler has to run *inside* the Docker containers, which use Linux.

- SmartBugs determines the correct compiler version by searching the source code for the first `pragma solidity` directive. In this example, it finds the line
  ```solidity
  pragma solidity >=0.6.6;
  ```
  SmartBugs loads the highest compiler version compatible with this directive. Well, actually not quite, since this would currently mean `0.8.29`. SmartBugs guesses that the programmer probably intended to write `^0.6.6` and therefore loads `0.6.12`.

- Semantic versioning requires that there are no breaking changes between `0.6.6` and `0.6.12` (upwards compatibility), so `0.6.12` should be fine. Unfortunately, there *have been* breaking changes: In the comments, the source code specifies several incompatible code licenses, which are ignored by `0.6.6` but raise a compile-time error with `0.6.12`.
  
  Note that picking the lowest compatible version, `0.6.6`, has its pitfalls, as people like to write `^0.6.0` when they are actually using a higher version, not bearing in mind that the compiler versions need not be backwards compatible.

**Fix 1:** Replace `>=0.6.6` by `0.6.6`. This tells SmartBugs to use this version and none higher. (But read on, better fix below.)

With this fix, the results are a bit better, but there are still compilation errors, for two reasons:

- Some old tools, like Oyente, Osiris and HoneyBadger, refuse to cooperate with compilers that were released after maintenance of these tools stopped.

- Tools like Mythril try to detect the Solidity version themselves, but do it wrongly. Mythril picks the first directive even though it is commented out:
  ```solidity
  // pragma solidity >=0.5.0;
  ```
  So Mythril expects a compiler for `0.5.0`, rejects the compiler `0.6.6` offered by SmartBugs, and fails.

**Fix 2 (subsumes Fix 1):** Add a new directive as the first line of the source code, instead of editing the directive further down.
```solidity
pragma solidity 0.6.0;
```
Now both, SmartBugs and Mythril, will go for this version.

**Fix 3:** You can get more results (sometimes even from old tools), if you also provide the deployment and runtime bytecodes of the contract for analysis. The bytecodes can be generated with the compiler, obtained from an archive node, or from websites like [Etherscan](https://etherscan.io/).

**Example:** Put [your source code](https://bscscan.com/address/0x7ac55ac530f2C29659573Bde0700c6758D69e677#code) into a file `DemaxPair.sol`. According to Etherscan, the main contract in the source file (i.e., the contract actually deployed at the address) is the one named `DemaxPair`, so we use it for the filename, thinking of the SmartBugs option `--main`. Download the version `0.6.6` of solc. Under Linux, the following commands store the bytecodes in `DemaxPair.hex` and `DemaxPair.rt.hex`, respectively.
```console
F=DemaxPair.sol
C=DemaxPair
solc --bin ${F} | awk 's==2{print;s=0}/=== '"$F"':'"$C"' ===/{s+=1}/Binary/&&s==1{s+=1}' > ${F/.sol/.hex}
solc --bin-runtime ${F} | awk 's==2{print;s=0}/=== '"$F"':'"$C"' ===/{s+=1}/Binary/&&s==1{s+=1}' > ${F/.sol/.rt.hex}
```
Now you can analyze all three versions of the contract, source and bytecodes, with the following command:
```console
./smartbugs -t all -f DemaxPair.* --timeout 300 --main
```
The option `--main` ensures that in the source code, only the contract named `DemaxPair` is analyzed, and the others are skipped. Some tools may run for a long time, so setting an adequate timeout (here 5 minutes) may be useful.

Regarding compilation errors, this is as good as it gets as long as the old tools aren't updated.

**Comment by JerryPW** [comment]
Thanks for your detailed explanation! It really helps me a lot!
Hopes your product better and better!
